### PR TITLE
Add introduction to the setting system

### PR DESCRIPTION
**Pull request by jsuereth**
Review by @eed3si9n or @havocp

### CODE REVIEW

File: src/sphinx/Architecture/Core-Principles.rst [hunk 1]
```diff
@@ -63,8 +63,63 @@ done through the ``Setting[_]`` sequence.
 Introduction to Settings
 ========================
 
-TODO - Discuss ``Setting[_]``
+A Setting represents the means of constructing the value of one particular ``AttributeKey[_]`` in the ``AttributeMap`` of build state.  A setting consists of two pieces:
```
Comment (havocp) [inline src/sphinx/Architecture/Core-Principles.rst:66]: I can imagine someone coming here looking for an end-user type introduction and they get this sentence and go "What... the... f...." - maybe retitle this page "Introduction to Settings Internals" or "How Settings Work Behind the Scenes" or "Settings: The Deep Dive" or something, and then at the top you could even link to the end-user intro to settings in the getting started guide... just worried about google ending up here when someone searches "introduction to settings"

File: src/sphinx/Architecture/Core-Principles.rst [hunk 2]
```diff
@@ -63,8 +63,63 @@ done through the ``Setting[_]`` sequence.
...
 
-TODO - Transition into ``Task[_]``
+1. The ``AttributeKey[T]`` where the value of the setting should be assigned.
+2. An ``Intialize[T]`` object which is able to construct the value for this setting.
 
-TODO - Transition into ``InputTask[_]``
\ No newline at end of file
+Sbt's intiialization time is basically just taking a sequence of these ``Setting[_]`` objects and running their initialization objects and then storing the value into the ``AttributeMap``.   This means overwriting an exisitng value at a key is as easy as appending a
```
Comment (havocp) [inline src/sphinx/Architecture/Core-Principles.rst]: human spell check:  intiialization exisitng

File: src/sphinx/Architecture/Core-Principles.rst [hunk 3]
```diff
@@ -63,8 +63,63 @@ done through the ``Setting[_]`` sequence.
...
+
+Where it gets interesting is that ``Initialize[T]`` can depend on other ``AttributeKey[_]``s in the build state.  Each ``Initialize[_]``
+can pull values from any ``AttributeKey[_]`` in the build state's ``AttributeMap`` to compute its value.  Sbt ensures a few things
+when it comes to ``Initialize[_]`` dependencies:
+
+1. There can be no circular dependencies
+2. If one ``Initialize[_]`` depends on another ``Initialize[_]`` key, then *all* associated ``Initialize[_]`` blocks for that key must
+   have run before we load the value.
+
+The above image shows a bit of what's expLet's look at what gets stored for the setting ::
```
Comment (havocp) [inline src/sphinx/Architecture/Core-Principles.rst]: sentence has issues

File: src/sphinx/Architecture/Core-Principles.rst [hunk 4]
```diff
@@ -63,8 +63,63 @@ done through the ``Setting[_]`` sequence.
...
+
+
+
+.. Note: This image comes from a google drawing: https://docs.google.com/a/typesafe.com/drawings/d/1hvE89XVrQiXdSBsgaQgQGTmcO44EBZPg4_0WxKXU7Pw/edit
+.. Feel free to request access to modify as appropriate.
+
+.. image:: overview-setting-example.png
+
+
+Here, a ``Setting[_]`` is constructed that understands it depends on the value in the ``name`` AttributeKey.  Its intiialize block first grabs the value of the ``name`` key, then runs the function normalize on it to compute its value.
```
Comment (havocp) [inline src/sphinx/Architecture/Core-Principles.rst]: your editor is now autocompleting intiialiaze or something.

File: src/sphinx/Architecture/Core-Principles.rst [hunk 5]
```diff
@@ -63,8 +63,63 @@ done through the ``Setting[_]`` sequence.
...
+
+
+Introduction to Tasks
+=====================
+
+The next layer in sbt is around these user request, or tasks.   When a user configures a build, they are defining a set of repeatable
+tasks that they can run on their project.  Things like ``compile`` or ``test``.   These tasks *also* have a dependency graph, where
+e.g. the ``test`` task requires that ``compile`` has run before it can successfully execute.
+
+Sbt's defines a class ``Task[T]``.  The ``T`` type parameter represents the type of data returned by a task.  Remember the tenenats of
```
Comment (havocp) [inline src/sphinx/Architecture/Core-Principles.rst]: tenets

File: src/sphinx/Architecture/Core-Principles.rst (hunk 1 above)
Comment (jsuereth) [inline src/sphinx/Architecture/Core-Principles.rst:66]: good point, corrected.

File: src/sphinx/Architecture/Core-Principles.rst (hunk 4 above)
Comment (jsuereth) [inline src/sphinx/Architecture/Core-Principles.rst]: don't have spellcheck installed in sublime yet...

**Comment by jsuereth** [comment]
Ok @havocp, fixed up the spelling issues.

**Pull request merged**